fix(ivy): support tsutils helper decorator definitions

petebacondarwin opened the pull request:
FW-379 #resolve 

**TODO:**
* [x] implement `__decorate` decorator removal

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

See FW-379


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mary-poppins commented:
You can preview 9db1e7f at https://pr25905-9db1e7f.ngbuilds.io/.

mary-poppins commented:
You can preview ab83153 at https://pr25905-ab83153.ngbuilds.io/.

mary-poppins commented:
You can preview 70513ef at https://pr25905-70513ef.ngbuilds.io/.

mary-poppins commented:
You can preview 61169e9 at https://pr25905-61169e9.ngbuilds.io/.

mary-poppins commented:
You can preview a733ac7 at https://pr25905-a733ac7.ngbuilds.io/.

mary-poppins commented:
You can preview 60f668e at https://pr25905-60f668e.ngbuilds.io/.

mhevery commented:
http://test/OCL:213329545:BASE:213329600:1537216884310:ebf8785e

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -131,7 +136,8 @@ export class Esm5ReflectionHost extends Fesm2015ReflectionHost {
 
   /**
-   * Constructors parameter decorators are declared in the body of static method of the constructor
+   * A constructor's parameter decorators are declared in the body of static method of the
```
well this is not grammatically correct - even though it is copied from before :-)

mary-poppins commented:
You can preview dd55f5d at https://pr25905-dd55f5d.ngbuilds.io/.

gkalpak reviewed (approved):
Some minor comments and typos. LGTM otherwise 👍

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -49,9 +50,9 @@ export class Esm5ReflectionHost extends Fesm2015ReflectionHost {
    *   outer variable.
    *
-   * @param node The top level declaration that represents an exported class or the function
+   * @param node the top level declaration that represents an exported class or the function
```
What is the convention? We have both cases throughout the code, so I was just trying to be consistent in this file.

gkalpak replied:
The convention is to not have a convention 😛 But capitalizing is a no brainer 😛
(I really don't understand why people don't use the `@param arg - Description` format. It's so much more readable :grin:)

petebacondarwin replied:
But this PR is not the place to rewrite all these `@param` tags :-)

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -64,22 +64,14 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+      return this.getClassDecoratorsFromStaticProperty(symbol);
+    } else {
+      return this.getClassDecoratorsFromHelperCall(symbol);
     }
```
The latter will just quietly return `null`, which is what we want. I decided it wasn't worth doing an extra check here.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -168,8 +203,24 @@ export class Esm5ReflectionHost extends Fesm2015ReflectionHost {
+   * @param classSymbol the class whose helper calls we are interested in.
+   * @returns an array of statements that may contain helper calls.
+   */
+  protected getStatementsForClass(classSymbol: ts.Symbol): ts.Statement[] {
```
Good question... `EsmReflectionHost.getClassSymbol()` is probably going to return the outer `FooClass`.
I will check.

petebacondarwin replied:
No unit tests for this (nor `getHelperCallsForClass`)!

petebacondarwin replied:
So `EsmReflectionHost.getClassSymbol()` actually does return the inner node, so all is well. I have added a check for this.

gkalpak replied:
:100:

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -211,8 +221,137 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+   * @param decoratorProperty the static property that would hold the decorators.
+   * @returns true if the class has the specified static property.
+   */
+  protected hasStaticPropertyDecorators(symbol: ts.Symbol, decoratorProperty: ts.__String):
```
+1

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/parsing/esm2015_parser.ts:
```diff
@@ -24,20 +24,24 @@ export class Esm2015FileParser implements FileParser {
-                const decorators = this.host.getDecoratorsOfDeclaration(declaration);
-                return decorators && declaration.name &&
-                    new ParsedClass(declaration.name.text, declaration, decorators);
+                if ((ts.isClassDeclaration(declaration) || ts.isVariableDeclaration(declaration))) {
```
+1

petebacondarwin committed:
fix(ivy): include "variable-declared" decorated classes in ngcc compilation

Previously, classes that were declared via variable declarations,
rather than class declarations, were being excluded from the
parsed classes.

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/host/esm5_host_import_helper_spec.ts:
```diff
@@ -0,0 +1,372 @@
+    });
+  });
+
+  const NGMODULE_FILE = {
```
+1

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/rendering/esm2015_renderer_spec.ts:
```diff
@@ -136,61 +174,121 @@ A.decorators = [
+         });
+
+
+      it('should delete the decorator (and its container if there are not other decorators left) that was matched in the analysis',
```
Grammatically it is correct, but your version reads better. :heart:

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -238,8 +382,143 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+          this.reflectDecoratorsFromHelperCall(helperCall, isMemberTarget(classSymbol.name));
+      decoratorMap.forEach((decorators, memberName) => {
+        if (memberName) {
+          const memberDecorators = memberDecoratorMap.get(memberName) || [];
```
Actually, that is not correct. The `memberDecoratorMap` is the output value not the one we are iterating over.

gkalpak replied:
Ooops 😇

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts:
```diff
@@ -74,6 +74,57 @@ function compileNgModuleFactory__POST_NGCC__(injector, options, moduleType) {
 export {A, B, C};`
 };
 
+const PROGRAM_DECORATE_HELPER = {
```
used - I have added the tests

petebacondarwin committed:
fix(ivy): ngcc: remove redundant `__decorate()` calls

Previously we only removed assignments to `Class.decorators = [];`
if the array was not empty.

Now we also remove calls to `__decorate([])`, similarly.

mary-poppins commented:
You can preview 08c6b1e at https://pr25905-08c6b1e.ngbuilds.io/.

gkalpak commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -221,10 +360,15 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+      Map<string, Decorator[]> {
     const memberDecorators = new Map<string, Decorator[]>();
-    if (classSymbol.exports && classSymbol.exports.has(PROP_DECORATORS)) {
+    if (classSymbol.exports) {
```
I should be more clear:
I meant, since you are already guarding the call to `getMemberDecoratorsFromStaticProperty()` with `hasStaticProperty()`, you can know that there is an `exports` property.
So, you could change the `hasStaticProperty()` signature to [return a type predicate][1]:

```diff
-protected hasStaticProperty(symbol: ts.Symbol, propertyName: ts.__String): boolean
+protected hasStaticProperty(symbol: ts.Symbol, propertyName: ts.__String): symbol is {exports: Map<ts.Symbol>}
```
[1]: http://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

gkalpak replied:
Actually,the correct return type is: `symbol is ts.Symbol & {exports: NonNullable<ts.Symbol['exports']>}` 👌

gkalpak replied:
This shouldn't be the case if you use `NonNullable` (see https://github.com/angular/angular/pull/25905#discussion_r220194878).

petebacondarwin replied:
In this case `[exports: NonNullable<ts.Symbol['exports']>}` and `{exports: ts.SymbolTable}` are the same type. That is not the problem. The problem is to find a way of telling TS that the static property on `exports` that we checked for is defined.

I.E.

```
symbol.exports.get('decorators') !== undefined
```

gkalpak replied:
I am talking about getting rid of the `if (classSymbol.exports)` check. `classSymbol.exports.get('decorators')` is another story.

petebacondarwin replied:
I went with this in the end: https://github.com/angular/angular/pull/25905/commits/9f6eac59ff2146567a212184b41480c5b33776f6

gkalpak replied:
BTW, if you really wanted TS to infer that `x.exports` has specific keys, you could do [something like that][1] :grin:

[1]: http://www.typescriptlang.org/play/index.html#src=%2F%2F%20Augment%20%60Map%23has()%60%20to%20make%20it%20aware%20of%20existing%20keys.%0D%0Ainterface%20Map%3CK%2C%20V%3E%20%7B%0D%0A%20%20has%3CT%20extends%20K%3E(key%3A%20T)%3A%20this%20is%20MapWithKeys%3CK%2C%20V%2C%20T%3E%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20MapWithKeys%3CK%2C%20V%2C%20Keys%20extends%20K%3E%20extends%20Map%3CK%2C%20V%3E%20%7B%0D%0A%20%20get(key%3A%20Keys)%3A%20V%3B%0D%0A%20%20get(key%3A%20K)%3A%20V%20%7C%20undefined%3B%0D%0A%0D%0A%20%20has%3CT%20extends%20Keys%3E(key%3A%20T)%3A%20true%3B%0D%0A%20%20has%3CT%20extends%20K%3E(key%3A%20T)%3A%20this%20is%20MapWithKeys%3CK%2C%20V%2C%20Keys%20%7C%20T%3E%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F%20Examples%0D%0Adeclare%20const%20m%3A%20Map%3Cstring%2C%20string%3E%3B%0D%0A%0D%0A%2F%2F%20We%20know%20%60v1%60%20and%20%60v2%60%20will%20se%20set%20to%20strings.%0D%0Aconst%20v1%3A%20string%20%3D%20m.has('foo')%20%3F%20m.get('foo')%20%3A%20'missing'%3B%0D%0Aconst%20v2%3A%20string%20%3D%20(m.has('foo')%20%26%26%20m.has('bar'))%20%3F%20m.get('bar')%20%3A%20'missing'%3B%0D%0A%0D%0A%2F%2F%20We%20don't%20know%20whether%20%60v3%60%20will%20be%20set%20to%20string%20or%20undefined.%20%0D%0Aconst%20v3%3A%20string%20%3D%20m.has('baz')%20%3F%20m.get('qux')%20%3A%20'missing'%3B%0D%0A%0D%0A%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%0D%0A%0D%0A%2F%2F%20Mock%20%60ts.Symbol%60.%0D%0Ainterface%20Symbol%20%7B%0D%0A%20%20exports%3F%3A%20Map%3Cstring%2C%20string%3E%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F%20Create%20an%20augmented%20%60ts.Symbol%60%20that%20is%20aware%20of%20existing%20exports.%0D%0Ainterface%20SymbolWithExports%3CKnownExports%20extends%20string%3E%20extends%20Symbol%20%7B%0D%0A%20%20exports%3A%20MapWithKeys%3Cstring%2C%20string%2C%20KnownExports%3E%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F%20Check%20if%20symbol%20has%20exports%20and%20type%20it%20accordingly.%0D%0Afunction%20hasStaticProperty%3CT%20extends%20string%3E(x%3A%20Symbol%2C%20prop%3A%20T)%3A%20x%20is%20SymbolWithExports%3CT%3E%20%7B%0D%0A%20%20return%20!!x.exports%20%26%26%20x.exports.has(prop)%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F%20Examples%0D%0Adeclare%20const%20x%3A%20Symbol%3B%0D%0Alet%20foo%3A%20string%3B%0D%0A%0D%0Aif%20(hasStaticProperty(x%2C%20'foo'))%20%7B%0D%0A%20%20%2F%2F%20We%20know%20%60foo%60%20exists.%20%0D%0A%20%20foo%20%3D%20x.exports.get('foo')%3B%0D%0A%7D%20else%20%7B%0D%0A%20%20%2F%2F%20We%20don't%20know%20whether%20%60foo%60%20exists.%0D%0A%20%20foo%20%3D%20x.exports!.get('foo')%3B%0D%0A%7D%0D%0A

petebacondarwin replied:
Right but life is not long enough :-)

alxhub reviewed (changes requested):
Nice job! Overall this looks very good. A few comments to address and deal with, nothing very major.

I also attempted to rebase your PR on master and the fixup commits didn't merge properly, so you might need to look into that to finally land this change. Feel free to rebase and apply them ahead of time.

mary-poppins commented:
You can preview 242a531 at https://pr25905-242a531.ngbuilds.io/.

mary-poppins commented:
You can preview e6d8b02 at https://pr25905-e6d8b02.ngbuilds.io/.

petebacondarwin committed:
fix(ivy): ngcc: only consider decorators from `@angular/core`

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/host/fesm2015_host_import_helper_spec.ts:
```diff
@@ -0,0 +1,426 @@
+  }
+];
+
+const INTERNAL_CORE_DECORATORS = [];
```
Ahah! That was the beginnings of a test for the internal imports of `@angular/core`.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -357,44 +636,209 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+  const importFrom = decorator.import && decorator.import.from || '';
+  return importFrom === '@angular/core' ||
+      (/^\./.test(importFrom) &&
+       /node_modules[\\\/]@angular[\\\/]core/.test(decorator.node.getSourceFile().fileName));
```
I need to add some unit tests for this too

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -211,8 +223,131 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
         [];
   }
 
+  getVariableValue(declaration: ts.VariableDeclaration): ts.Expression|undefined {
```
TODO...

petebacondarwin replied:
So I am going to suggest that we check that the value being assigned is the return value from a call to `__decorate()`. This should be pretty constrained.

petebacondarwin replied:
Done

mary-poppins commented:
You can preview c82e1fc at https://pr25905-c82e1fc.ngbuilds.io/.

mary-poppins commented:
You can preview a6f4743 at https://pr25905-a6f4743.ngbuilds.io/.

petebacondarwin commented:
@alxhub - PTAL. I have rebased (and squashed previous fixup commits). I have addressed all your comments in new commits (some are marked as fixups). The new commits begin from https://github.com/angular/angular/pull/25905/commits/6c94c1a5548d1e5c1934984dd554a62b4e71ebb8.

The two main areas that need a good review are:

* how core imports are determined when the package is already `@angular/core`: https://github.com/angular/angular/pull/25905/commits/c9e1f8f31459b6413bfb101681d3738b04bf9cfe
* the new approach to finding the value of a late initialized variable: https://github.com/angular/angular/pull/25905/commits/a6f4743b24c70c65cf2278c8026be889daf38083

mary-poppins commented:
You can preview e117114 at https://pr25905-e117114.ngbuilds.io/.

mary-poppins commented:
You can preview f9f11a7 at https://pr25905-f9f11a7.ngbuilds.io/.

petebacondarwin committed:
fix(ivy): support `tsutils.__decorate` decorator declarations in ngcc

The most recent Angular distributions have begun to use __decorate instead of Class.decorators.
This prevents `ngcc` from recognizing the classes and then fails to perform the transform to
ivy format.

Example:

```
var ApplicationModule = /** @class */ (function () {
    // Inject ApplicationRef to make it eager...
    function ApplicationModule(appRef) {
    }
    ApplicationModule = __decorate([
        NgModule({ providers: APPLICATION_MODULE_PROVIDERS }),
        __metadata("design:paramtypes", [ApplicationRef])
    ], ApplicationModule);
    return ApplicationModule;
}());
```

Now `ngcc` recognizes `__decorate([...])` declarations and performs its transform.

See FW-379

mary-poppins commented:
You can preview 0b32b9e at https://pr25905-0b32b9e.ngbuilds.io/.

mary-poppins commented:
You can preview 6dbb14c at https://pr25905-6dbb14c.ngbuilds.io/.

petebacondarwin commented:
@alxhub - PTAL. I have rebased on master (and squashed all fixup commits).
I have addressed all your comments but I had a terrible time getting all the fix up commits to work together so I have cleaned them all out.

The two main areas that need a good review are:

* how core imports are determined when the package is already `@angular/core`: https://github.com/angular/angular/pull/25905/commits/6dbb14ca8115955026b0f4858eb8c68c3d8b9194
* the new approach to finding the value of a late initialized variable: https://github.com/angular/angular/pull/25905/commits/14621bdc24ab387e95f2a4686767db73a80452ba

gkalpak commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -357,44 +697,270 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+      return null;
+    }
+
+    const importClause = namespaceDeclaration && ts.isNamespaceImport(namespaceDeclaration) ?
```
Nit: The `namespaceDeclaration && ts.isNamespaceImport()` shouldn't be necessary here. At this point, `namespaceDeclaration` is guaranteed to be of type `ts.NamespaceImport` (and TS should be able to infer that).

gkalpak commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -357,44 +697,270 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+      if (!importInfo) {
+        return {name: identifier.text, from: this.packageName};
+      }
+      if (/^\./.test(importInfo.from)) {
```
So, all relative imports are marked as `@angular/core`? Why is that?

petebacondarwin replied:
But basically, no this is wrong because it breaks the renderer, which tries to add `import ... from `@angular/core` even when we are already inside `@angular/core`.

petebacondarwin committed:
fix(ivy): support late-initialized variables in ngcc/ngtsc

In some formats variables are declared as `var` or `let` and only
assigned a value later in the code.

The ngtsc resolver still needs to be able to resolve this value,
so the host now provides a `host.getVariableValue(declaration)`
method that can do this resolution based on the format.

The hosts make some assumptions about the layout of the
code, so they may only work in the constrained scenarios that
ngcc expects.

mary-poppins commented:
You can preview 65a7c22 at https://pr25905-65a7c22.ngbuilds.io/.

gkalpak commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -211,8 +221,186 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+      if (ts.isIdentifier(left) && this.checker.getSymbolAtLocation(left) === symbol) {
+        return (ts.isCallExpression(right) && getCalleeName(right) === '__decorate') ? right : null;
+      }
+      return this.findDecoratedVariableValue(right, symbol);
```
You are right :grin:

mary-poppins commented:
You can preview e1b46c3 at https://pr25905-e1b46c3.ngbuilds.io/.

petebacondarwin commented:
To simplify matters I have removed some of the more recent commits from this PR. This way we can land this and then worry about the other fixes in other PRs.

@alxhub PTAL

mary-poppins commented:
You can preview 4ae2603 at https://pr25905-4ae2603.ngbuilds.io/.

mary-poppins commented:
You can preview 9679f96 at https://pr25905-9679f96.ngbuilds.io/.

mary-poppins commented:
You can preview d7bb57c at https://pr25905-d7bb57c.ngbuilds.io/.

petebacondarwin commented:
Superceded by https://github.com/angular/angular/pull/26236
